### PR TITLE
Fix cases where too much vectorization was applied in fusion

### PR DESCRIPTION
**Pull request by krzysz00**
There exist some fusions (such as broadcasting a scalar) where we could get bad results where the vectorization of the gemm writeback is larger than the vectorization that the fusion argument (which is transformed to give it the same shape as the gemm output).

Address this bug by having the load loop for additional arguments be vectorized by a vectorization that's a function of the fusion-argument transforms and capped at the vectorization of the writeback. This will almost certainly lead to scalar loads.

Furthermore, resolve a long-standing bug in index diff updates that could lead to use-before-def but somehow didn't trigger until this bug fix.

@sjw36 Can you get me the test case that led you to find this issue so we can add it in here?

### CODE REVIEW

File: mlir/lib/Dialect/Rock/Transforms/AlignTiling.cpp [hunk 1]
```diff
@@ -217,29 +215,50 @@ static void insertLoadFromOtherSource(PatternRewriter &b, Location loc,
...
+
+    Type typeToLoad = destElemType;
+    if (maxVectorLen > 1)
+      typeToLoad = VectorType::get(maxVectorLen, typeToLoad);
+
     auto copyLoop = b.create<TransformingForOp>(
-        loc, ArrayRef<ValueRange>{loadCoord},
-        ArrayRef<Attribute>{sourceTransformsFromOp}, /*bounds=*/consts,
-        /*strides=*/ArrayRef<int64_t>(consts), /*forceUnroll=*/true,
+        loc, ArrayRef<ValueRange>{loadCoord, zeroes},
```
Comment (jungpark-mlir) [inline mlir/lib/Dialect/Rock/Transforms/AlignTiling.cpp:249]: Was this entirely wrong or just the given vector width doesn't align?
I feel this solution lowers the transform in early stage.

Is there any way to handle it in lowering transformingFor?

Comment (krzysz00) [reply mlir/lib/Dialect/Rock/Transforms/AlignTiling.cpp:249]: This isn't lowering anything?

This just adds a second domain that gives us an actual loop index to store at.

**Comment by sjw36** [comment]
@krzysz00 I have validated that the testcase passes with this change but I think it will need my changes as well. https://github.com/ROCmSoftwarePlatform/llvm-project-private/issues/727

**Review by sjw36 — APPROVED** [review]
Fixes bug. Thanks!

File: mlir/lib/Dialect/Rock/Transforms/AlignTiling.cpp (hunk 1 above)
Comment (jungpark-mlir) [reply mlir/lib/Dialect/Rock/Transforms/AlignTiling.cpp:249]: Right, I see.

**Review by jungpark-mlir — APPROVED** [review]

**Pull request merged**